### ISSUE TITLE
softmax and cross-entropy VJPs violate logit-translation invariance in float64

### ISSUE THREAD

**Issue by beanduan22** [tensorflow/tensorflow#126525]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

TensorFlow version: 2.21.0

### Custom code

Yes

### Python version

Python version: 3.13.5

### Current behavior?

The gradients returned by TensorFlow for `tf.nn.softmax` and `tf.nn.sparse_softmax_cross_entropy_with_logits` have a nonzero sum at a finite float64 input, despite both forward functions being invariant to adding the same scalar to every logit.

```text
softmax(z + c * 1) = softmax(z)
cross_entropy(z + c * 1, target) = cross_entropy(z, target)
```

Therefore their logit gradients must sum to zero. At `z = [40.0, 0.0]`, the actual residual has magnitude `4.248354255291589e-18`, a finite normal float64 value.

### Standalone code to reproduce the issue

```shell
import math
import tensorflow as tf

def run(device):
    with tf.device(device):
        logits = tf.Variable([40.0, 0.0], dtype=tf.float64)
        with tf.GradientTape() as tape:
            output = tf.nn.softmax(logits)[0]
        softmax_grad = tape.gradient(output, logits)
        logits = tf.Variable([[40.0, 0.0]], dtype=tf.float64)
        with tf.GradientTape() as tape:
            loss = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=[0], logits=logits)[0]
        cross_entropy_grad = tape.gradient(loss, logits)
    print(device)
    print("softmax VJP:", softmax_grad.numpy().tolist())
    print("softmax VJP sum:", float(tf.reduce_sum(softmax_grad)))
    print("cross entropy VJP:", cross_entropy_grad.numpy()[0].tolist())
    print("cross entropy VJP sum:", float(tf.reduce_sum(cross_entropy_grad)))

run("/CPU:0")
if tf.config.list_physical_devices("GPU"):
    run("/GPU:0")
print("expected tail magnitude:", math.exp(-40.0) / (1.0 + math.exp(-40.0)) ** 2)
```

### Relevant log output

```shell
## Actual result

On TensorFlow `2.21.0`, CPU and GPU both produce:


softmax VJP: [0.0, -4.248354255291589e-18]
softmax VJP sum: -4.248354255291589e-18
cross entropy VJP: [0.0, 4.248354255291589e-18]
cross entropy VJP sum: 4.248354255291589e-18
expected tail magnitude: 4.248354255291589e-18


## Expected result

Both VJPs should sum to zero. The expected softmax VJP is approximately `[+4.248354255291589e-18, -4.248354255291589e-18]`; the cross-entropy VJP is its negation.
```

**Comment by animesh68** [comment]
Hi @Venkat6871 , can I work on this issue? I’ve reproduced it locally and would be happy to take it up.

**Comment by Venkat6871** [comment]
Hi @animesh68 ,
Yes, absolutely! TensorFlow is open source, and we welcome contributions from the community.

Please follow the official TensorFlow contribution guidelines to get started. I am adding the [documentation link](https://www.tensorflow.org/community/contribute/code) here for reference.

Thank you for your interest in working on this issue!

**Comment by animesh68** [comment]
Thanks @Venkat6871! I’ll get started on this issue and follow the TensorFlow contribution guidelines. I’ll investigate the VJP implementations for `tf.nn.softmax` and `tf.nn.sparse_softmax_cross_entropy_with_logits`, add appropriate regression tests, and submit a PR once I have a fix.

**Comment by Kayyuri** [comment]
Hi @beanduan22, Thank you for reporting this issue!
A fix for this issue has been proposed in this PR(https://github.com/tensorflow/tensorflow/pull/126748). Once it is merged, this issue will be resolved.

Many thanks to @AshiteshSingh for raising the PR to address this issue.
